pandas/_libs/index.pyx

```diff
@@ -606,25 +606,107 @@ cdef class BaseMultiIndexCodesEngine:
-            indexer = (getattr(self._base, f'get_{method}_indexer')
-                       (self, lab_ints, limit=limit))
-            indexer = indexer[order]
-        else:
-            indexer = self._base.get_indexer(self, lab_ints)
+    def get_indexer_and_fill(self, object values, object target,
+                             object method, object limit = None) -> np.ndarray:
+        """
+        Gets an indexer, i.e. a set of indexes into `values`, for the values in
+        `target`, where the index value.
 
-        return indexer
+        If method is "backfill" then the index for a value in `target` which
+        does not exist in `values` is the index of the next match, or -1 is the
+        value is larger than the largest value in `values`.
+
+        Similarly, if the method if "pad" then the index for a value in `target`
+        which does not exist in `values` is the index of the previous match, or
+        -1 if the value is smaller then the largest value in `values`.
+
+        Parameters
+        ----------
+        values : list-like of tuples
+            must be sorted and all have the same length
+        target: list-like of tuples
+            need not be sorted, but all must have the same length, which must be
+            the same as the length of all tuples in `values`
+        method: string
+            "backfill" or "pad"
+        limit: int, optional
+            if provided, limit the number of fills to this value
+
+        Returns
+        -------
+        np.ndarray[int64_t, ndim=1] of the indexer of `target` into `values`,
+        filled with the `method` (and optionally `limit`) specified
+        """
+        assert method in ("backfill", "pad")
+        cdef:
+            int64_t i, j, next_code
+            int64_t num_values, num_target_values
+            ndarray[int64_t, ndim=1] target_order
+            ndarray[object, ndim=1] target_values
+            ndarray[int64_t, ndim=1] new_codes, new_target_codes
+            ndarray[int64_t, ndim=1] sorted_indexer
+
+        target_order = np.argsort(target.values).astype('int64')
+        target_values = target.values[target_order]
+        num_values, num_target_values = len(values), len(target_values)
+        new_codes, new_target_codes = (
+            np.empty((num_values,)).astype('int64'),
+            np.empty((num_target_values,)).astype('int64'),
+        )
+
+        # `values` and `target_values` are both sorted, so we walk through them
+        # and memoize the (ordered) set of indices in the (implicit) merged-and
+        # sorted list of the two which belong to each of them
+        # the effect of this is to create a factorization for the (sorted)
+        # merger of the index values, where `new_codes` and `new_target_codes`
+        # are the subset of the factors which appear in `values` and `target`,
+        # respectively
+        i, j, next_code = 0, 0, 0
+        while i < num_values and j < num_target_values:
```

Comment:
in python-space a while loop is slower than a for loop.  no idea if it matters in cython

Reply:
As far as I can tell, it's just a question of how it compiles.  I'd guess that a for-loop with `break` statements would compile similarly, and that the only difference here would be the location of the `jne` statements.  If you'd like, though, I can either profile it and/or send you a diff of the relevant `objdump` snippets?